src/progress_bar.rs

```diff
@@ -511,6 +511,30 @@ impl ProgressBar {
         }
     }
 
+    /// Wraps a [`futures::Stream`](https://docs.rs/futures/0.3/futures/stream/trait.StreamExt.html) with the progress bar
+    ///
+    /// ```
+    /// # use indicatif::ProgressBar;
+    /// use futures::stream::{self, StreamExt};
+    /// # async fn test() {
+    /// let pb = ProgressBar::new(10);
+    ///
+    /// let stream = pb.wrap_stream(stream::iter(1..=10));
+    /// assert_eq!(
+    ///     stream.count().await,
+    ///     10,
+    /// );
+    /// # }
+    /// ```
+    #[cfg(feature = "futures")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "futures")))]
+    pub fn wrap_stream<S: futures_core::Stream>(&self, stream: S) -> ProgressBarIter<S> {
```

Comment:
Style nit: would prefer `impl futures_core::Stream` to avoid the indirection to a type argument here.

Reply:
I've applied this change, but think I will revert. As the code currently stands, we lose the type of the streamed item.
```rust
    pub fn wrap_stream(
        &self,
        stream: impl futures_core::Stream,
    ) -> ProgressBarIter<impl futures_core::Stream> {
        ProgressBarIter {
            progress: self.clone(),
            it: stream,
        }
    }

```

We've not told `rustc` that the item is the same, so this code doesn't compile:
```rust
fn main() {
    use futures::StreamExt;
    let collected = futures::executor::block_on(
        indicatif::ProgressBar::new(100)
            .wrap_stream(futures::stream::iter(0..100))
            .inspect(|_| std::thread::sleep(std::time::Duration::from_millis(100)))
            .collect::<Vec<usize>>(),
    );
    println!("{collected:?}")
}
```
```
error[E0277]: the trait bound `std::vec::Vec<usize>: std::iter::Extend<<impl futures::Stream as futures::Stream>::Item>` is not satisfied
   --> examples/stream.rs:4:9
    |
3   |       let collected = futures::executor::block_on(
    |                       --------------------------- required by a bound introduced by this call
4   | /         indicatif::ProgressBar::new(100)
5   | |             .wrap_stream(futures::stream::iter(0..100))
6   | |             .inspect(|_| std::thread::sleep(std::time::Duration::from_millis(100)))
7   | |             .collect::<Vec<usize>>(),
    | |____________________________________^ the trait `std::iter::Extend<<impl futures::Stream as futures::Stream>::Item>` is not implemented for `std::vec::Vec<usize>`
    |
```

So we'd at least need to write this:
```rust

    pub fn wrap_stream<I>(
        &self,
        stream: impl futures_core::Stream<Item = I>,
    ) -> ProgressBarIter<impl futures_core::Stream<Item = I>> {
```

Which means we might as well keep the `<S: futures_core::Stream>` bound.

In addition, not erasing the type might be important - maybe the `impl Stream` is also `impl Send`, but we force `rustc` to forget that fact.

Reply:
Ahhh sorry, yes, because we're reusing the argument type for the return type this doesn't work.